Reactの基本の提出物

ham-cap opened the pull request:
ReactでToDoアプリを作成いたしました。
よろしくお願いいたします。

ham-cap committed:
CDNでReactとBabelを読み込み・todo.jsを読み込み

ham-cap committed:
CreateForm、EditForm、TodoListコンポーネントを作成（要素の配置のみで機能は未実装）

ham-cap committed:
ESLintを設定

ham-cap committed:
新規登録機能を実装

ham-cap committed:
ESLintのno-undefルールを無効化

・JSXの構文内においては既にReactの宣言（import）は不要となっているが、ESLintから指摘され続け、視覚的にノイズになるためオフにした。

ham-cap committed:
新規登録機能を修正

・これまではCreateFormのstateにnewTodoとしてユーザーの入力した文字列を保持し、フォームの送信時にTodoAppのaddTodo関数へ渡していたが、CreateFormがそれを保持する必要がないため、newTodoをTodoAppのstateとして配置し直した。

ham-cap committed:
TodoListをクラスコンポーネントに変更

・現状、関数コンポーネントにしている意味は特にないため、親であるTodoAppコンポーネントに倣ってクラスコンポーネントに統一した。

ham-cap committed:
EditFormをクラスコンポーネントへ変更・ユーザーの入力を常にTodoAppのstateへ反映するよう変更

・TodoListと同様に、関数コンポーネントである必要がなかったため、TodoAppと統一した。
・編集機能を実装する前段階として、親であるTodoAppのstateとバインディングした。

ham-cap committed:
削除機能を実装

ham-cap committed:
ESLintとPrettierのルールが競合したため.eslintrcにeslint-config-prettierの記述を追加

ham-cap committed:
編集機能を実装

ham-cap committed:
新規登録機能の修正

・編集機能実装の際に入力フォームを切り替えるFormSwitcherを追加したことで、コンポーネント間の親子関係が変わり、そのままでは新規登録機能が正常に動作しないため、propsの受け渡しを修正した。

ham-cap committed:
編集機能を修正・編集フォームにキャンセルボタンを設置

・編集後に新規登録フォームへの切り替えができていなかったため修正した。
・編集しない場合に新規登録へ戻れるようにボタンを設置した。

ham-cap committed:
FormSwitcherの記述位置を変更

・可読性の観点から、コンポーネントの親子関係を把握しやすくするために親から順に記述することとした。

ham-cap committed:
Prettierで整形・labelタグ内の使用していない属性を削除・編集フォームのlabelの文言を修正

yoshitsugu commented on todo.js:
```diff
@@ -1 +1,205 @@
+  }
+
+  getTodosFromLocalStrage() {
+    if (Object.prototype.hasOwnProperty.call(localStorage, "todoList")) {
```
localStorageのgetItemはなければnullを返すことが仕様上決まっているので、hasOwnProperty使わずに、getItemしてみてnullだったら空の配列をいれる、でいいと思いますよ
https://html.spec.whatwg.org/multipage/webstorage.html#storage

yoshitsugu commented on todo.js:
```diff
@@ -1 +1,205 @@
+class TodoApp extends React.Component {
+  constructor(props) {
+    super(props);
+    const currentTodos = this.getTodosFromLocalStrage();
```
constructorで準備するとタイミング的にstateが準備されているのが保障されるのかどうかちょっと怪しいので、componentDidMountを使うほうが無難かなーと思います。
https://react.dev/reference/react/Component#componentdidmount
↑のドキュメントの通り、localStorageに限らず、データのフェッチのタイミングとしてよく使われます。

yoshitsugu reviewed (commented):
おおむねよさそうです。
コメントしました。

ham-cap committed:
getTodosFromLocalStrageの処理を修正・currentTodosのデータをフェッチするタイミングを変更

・getTodosFromLocalStrageではhasOwnPropertyを使用してlocalStorageにtodoListが存在しているかどうかをチェックし、その結果によって処理を分岐していたが、localStorage.getItemはアイテムが存在していない場合はnullを返す仕様のため、getItemの結果がnullか否かだけで存在チェックができ、記述も簡略化できるため変更した。

・localStorageに保存されている既存データはTodoAppのconstructorでstateを定義する直前にフェッチしていたが、このタイミングではstateの定義が完了している保証
がないため、確実にstateの準備ができているマウント直後にフェッチするよう変更した。

ham-cap commented:
修正いたしました！
よろしくお願いいたします！

yoshitsugu commented on todo.js:
```diff
@@ -1 +1,209 @@
+  }
+
+  getTodosFromLocalStrage() {
+    if (localStorage.getItem("todoList") === null) {
+      return [];
+    } else {
+      const todoListJson = localStorage.getItem("todoList");
+      const todoList = JSON.parse(todoListJson);
+      return todoList;
+    }
```
`else` の場合に2回 `getItem` することになっちゃうので、

```js
      const todoListJson = localStorage.getItem("todoList");
      if (todoListJson === null) {
        return [];
      }
      const todoList = JSON.parse(todoListJson);
      return todoList;
```

でよいかなと思います。

ham-cap replied:
ほんとですね💧
修正します！

yoshitsugu reviewed (commented):
一点だけコメントしました。

ham-cap committed:
getTodosFromLocalStrageを修正

・条件式にlocalStorage.getItem("todoList")を入れてしまうとelseの場合にgetItemが2回実行されてしまうため1回で済むように修正した。

ham-cap commented:
修正いたしました🙏

yoshitsugu reviewed (approved):
OKです～
